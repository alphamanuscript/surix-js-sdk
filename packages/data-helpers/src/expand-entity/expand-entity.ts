import { DataField, Entity, RawEntity, RawEntityData } from '../types';

function expandEntity (entity: Entity): RawEntity {
    const { data } = entity;
    const convertedData = expandEntityData(data);
    // delete entity.data;
    // @ts-ignore
    return { ...entity, data: convertedData };
}

function expandEntityData (data: DataField): RawEntityData {
    const expandedEntity = Object.keys(data)
        .reduce((expanded, key) => {
        let converted = {};
        switch(typeof data[key]) {
            case 'object':
                if(Array.isArray(data[key])) {
                    converted = {
                        type: 'list',
                        // @ts-ignore
                        value: data[key].map(item => expandEntityData(item))
                    };
                } else if (data[key] instanceof Date) {
                    converted = {
                        type: 'datetime',
                        value: data[key]
                    };
                } else {
                    converted = {
                        type: 'object',
                        // @ts-ignore
                        value: expandEntityData(data[key])
                    }
                }
                break;
            case 'string':
                converted = {
                    type: 'text',
                    value: data[key]
                };
            break;
            default:
                converted = { 
                    type: typeof data[key],
                    value: data[key]
                };
            break;
        }
        return { ...expanded, [key]: converted };
        }, {});
    return expandedEntity;
  }

export {
    expandEntity
};