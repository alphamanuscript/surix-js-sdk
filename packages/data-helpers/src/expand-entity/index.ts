export { expandEntity, expandArrayValue as expandArray } from './expand-entity';