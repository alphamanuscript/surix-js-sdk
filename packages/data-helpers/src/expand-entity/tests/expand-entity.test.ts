import { expandEntity } from '../expand-entity';

describe('ExpandEntity', () => {
    let entity;

    beforeAll(async () => {
        entity = {
            data: {
                name: 'Kevin Nderitu',
                age: 12,
                address: {
                    location: 'Nairobi'
                }
            }
        };
    });
    it('should be a function', async () => {
        expect(expandEntity).toBeInstanceOf(Function);
    });
    it('should convert successfully', async () => {
        const expected = {
            data: {
                name: {
                    type: 'text',
                    value: entity.data.name
                },
                age: {
                    type: 'number',
                    value: entity.data.age
                },
                address: {
                    type: 'object',
                    value: {
                        location: {
                            type: 'text',
                            value: entity.data.address.location
                        }
                    }
                }
            }
        };
        expect(expandEntity(entity)).toEqual(expected)
    });
});