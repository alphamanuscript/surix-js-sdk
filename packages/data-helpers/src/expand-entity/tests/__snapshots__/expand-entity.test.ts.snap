// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpandEntity should convert successfully 1`] = `
Object {
  "_id": "thisisanid",
  "data": Object {
    "address": Object {
      "type": "object",
      "value": Object {
        "location": Object {
          "type": "text",
          "value": "Nairobi",
        },
      },
    },
    "age": Object {
      "type": "number",
      "value": 12,
    },
    "birthDate": Object {
      "type": "datetime",
      "value": 1998-12-31T21:00:00.000Z,
    },
    "children": Object {
      "type": "list",
      "value": Array [
        Object {
          "name": Object {
            "type": "text",
            "value": "First son",
          },
        },
        Object {
          "name": Object {
            "type": "text",
            "value": "Second daughter",
          },
        },
      ],
    },
    "isOldEnough": Object {
      "type": "boolean",
      "value": true,
    },
    "name": Object {
      "type": "text",
      "value": "Kevin Nderitu",
    },
  },
  "tags": Array [
    "sometag",
  ],
}
`;
